Added Custom status emoji in rhs posts

manojmalik20 opened the pull request:
Modified the getCurrentUserLastPostGroupFirstPostId util to handle comments and replies

<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
A description of what this pull request does.
-->

#### Ticket Link
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/mattermost/mattermost-server/issues/XXXXX

Otherwise, link the JIRA ticket.
-->

#### Related Pull Requests
<!--
List all PRs related to resolving a ticket. For instance, if you submitted a PR to `mattermost/mattermost-server`, please include it here.
-->
- Has server changes (please link here)
- Has redux changes (please link here)
- Has mobile changes (please link here)

#### Screenshots
<!--
If the PR includes UI changes, include screenshots/GIFs.
-->

manojmalik20 committed:
Added Custom status emoji in rhs posts
Modified the getCurrentUserLastPostGroupFirstPostId util to handle comments and replies

manojmalik20 committed:
Review fixes
Moved Enable custom status config setting to last of the page
Modified custom status modal to show only those suggestions that are not in the recent list
Fixed CSS of the custom status modal

chetanyakan commented on components/custom_status/custom_status.scss:
```diff
@@ -1,6 +1,6 @@
-    //to override the styles from GenericModal class
-    width: 600px !important;
+    width: 600px;
+    //to override the styles from GenericModal and modal-dialog class
```
Remove this comment

chetanyakan commented on components/custom_status/custom_status_modal.tsx:
```diff
@@ -170,39 +178,42 @@ const CustomStatusModal: React.FC<Props> = (props: Props) => {
 
+    const renderCustomStatusSuggestions = () => {
+        const recentCustomStatusTexts = recentCustomStatuses.map((status: any) => status.text);
+        const customStatusSuggestions: any[] = [];
```
Use the appropriate type instead of `any`

chetanyakan commented on components/custom_status/custom_status_modal.tsx:
```diff
@@ -28,6 +29,13 @@ type Props = {
 };
 
 const EMOJI_PICKER_WIDTH_OFFSET = 308;
+const suggestedCustomStatuses: UserCustomStatus[] = [
```
change name

```suggestion
const defaultCustomStatusSuggestions: UserCustomStatus[] = [
```

chetanyakan commented on components/custom_status/custom_status_modal.tsx:
```diff
@@ -170,39 +178,42 @@ const CustomStatusModal: React.FC<Props> = (props: Props) => {
 
+    const renderCustomStatusSuggestions = () => {
+        const recentCustomStatusTexts = recentCustomStatuses.map((status: any) => status.text);
+        const customStatusSuggestions: any[] = [];
+        suggestedCustomStatuses.map((status: UserCustomStatus) => {
+            if (!recentCustomStatusTexts.includes(status.text)) {
+                customStatusSuggestions.push(
+                    <CustomStatusSuggestion
+                        handleSuggestionClick={handleSuggestionClick}
+                        emoji={status.emoji}
+                        text={status.text}
+                    />,
+                );
+            }
+            return null;
+        });
```
```suggestion
        const customStatusSuggestions = defaultCustomStatusSuggestions
            .filter((status: UserCustomStatus) => !recentCustomStatusTexts.includes(status.text))
            .map((status: UserCustomStatus) => (
                    <CustomStatusSuggestion
                        handleSuggestionClick={handleSuggestionClick}
                        emoji={status.emoji}
                        text={status.text}
                    />,
                ));
```

manojmalik20 replied:
In your suggested approach we are iterating the array twice... isn't is better if we iterate only once?

chetanyakan replied:
1. Some browsers will optimize the chaining of operators, (so it runs even faster than non-chaining code).
2. Iterating an array of length 5 once instead of twice is not that much optimization.
3. The code is much more readable this way.

manojmalik20 committed:
Review and post rebase fixes
Optimized the post list not to find currentUserLastPost if Update status post header is not visible
Fixed the styling of the channel header custom status
Fixed the coloring of the Set status menu option in status dropdown
Fixed the styling of the custom status modal
Fixed the styling of the custom status tooltip everywhere
Fixed the styling of the Update your status post header
Added custom status emoji in the sidebar channel where new component is used after rebase
Fixed the styling of the new emoji added in the sidebar channel list

manojmalik20 committed:
Typecheck fixes and updated snapshots

chetanyakan commented on utils/post_utils.jsx:
```diff
@@ -511,31 +512,50 @@ export function getValidUserPostFromId(state, id) {
+        return false;
+    }
+
+    //If current post and previous posts both are not comments
+    if (!post.root_id && !previousPost.root_id) {
+        //If a non-comment post has some replies then its header must be visible
+        //If a non-comment post is a consecutive post then its header must be hidden
+        return getReplyCount(post) > 0 || !areConsecutivePostsBySameUser(post, previousPost);
+    }
+
+    return true;
```
If this condition is still correct, can we simplify the code and comments like this:

```suggestion
    //If the current post or the previous posts is a comment
    if (post.root_id || previousPost.root_id) {
        return true;
    }
    
    //else If the current post has some replies then its header must be visible
    if (getReplyCount(post) > 0) {
        return true;
    }
    
    //else If the current post and previous post are consequetive, then the current post's header must be hidden
    if (areConsecutivePostsBySameUser(post, previousPost)) {
        return false;
    }

    return true;
```

chetanyakan commented on utils/post_utils.jsx:
```diff
@@ -511,31 +512,50 @@ export function getValidUserPostFromId(state, id) {
+    return true;
+}
+
 export function getCurrentUserLastPostGroupFirstPostId(state, postListIds) {
```
This name isn't correct anymore as replies break the post groups.

Rename to - `getLastCurrentUserPostWithHeaderId`

chetanyakan commented on utils/post_utils.jsx:
```diff
@@ -511,31 +512,50 @@ export function getValidUserPostFromId(state, id) {
-        } else if (currentUserLastPost) {
-            return currentUserLastPost.id;
+        if (isPostOwner(state, post) && !isSystemMessage(post) && isPostHeaderVisible(post, nextPost, getReplyCountModified)) {
+            return post.id;
```
Check if something like this would work:

```ts
// This function returns the postID of the last current user's post which has a header visible
export function getLastCurrentUserPostWithHeaderId(state, postListIds) {
        const getReplyCount = makeGetReplyCount().bind(null, state);

	const firstPostWithHeaderID = postListIds.find((postID, index) => {
		const post = getValidUserPostFromId(state, postID);
		
		let nextPost;
		if (index + 1 < postListIds.length) {
			const nextPostID = postListIds[index + 1];
			nextPost = getValidUserPostFromId(state, nextPostID);
		}

		return (isPostOwner(state, post) && !isSystemMessage(post) && isPostHeaderVisible(post, nextPost, getReplyCount));
	});

	return firstPostWithHeaderID || '';
}
```

manojmalik20 replied:
In this code, every time getReplyCount is called, makeGetReplyCount() is called. Isn't is better if makeGetReplyCount() is called only once just like it is currently being called.

manojmalik20 replied:
The find approach works as expected but the only drawback is that we have to find two posts from ids in every iteration.... but in the for loop approach we don't have to find nextPost if it's already present.

chetanyakan replied:
1. @manojosh I've updated the original post as well. I think it's better if we use:

```ts
const getReplyCount = makeGetReplyCount().bind(null, state);
```

2. Calling `getValidUserPostFromId` once or twice is still the same order of operations. So, there is not much optimization. We should prioritize readability and maintainability in this case.

manojmalik20 committed:
Modified the getLastCurrentUserPostWithHeaderId util for better readability
